File: Data Structures & Algorithms/subsets-ii/submission-1.java
class Solution {

    private void solve(int index,
                       int[] nums,
                       List<Integer> tmp,
                       List<List<Integer>> ans) {

        if (index == nums.length) {
            ans.add(new ArrayList<>(tmp));
            return;
        }

        tmp.add(nums[index]);
        solve(index + 1,
              nums,
              tmp,
              ans);

        tmp.remove(tmp.size() - 1);

        while (index + 1 < nums.length &&
               nums[index] == nums[index + 1]) {

            index++;
        }

        solve(index + 1,
              nums,
              tmp,
              ans);
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {

        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();

        solve(0,
              nums,
              new ArrayList<>(),
              ans);

        return ans;
    }
}

File: Data Structures & Algorithms/valid-palindrome-ii/submission-1.cpp
class Solution {
   public:
    bool solve(int l, int r, string& s) {
        while (l < r) {
            if (s[l] != s[r]) return false;
            l++;
            r--;
        }
        return true;
    }

    bool validPalindrome(string s) {
        int cnt = 0;

        int l = 0, r = s.size() - 1;

        while (l < r) {
            if (s[l] != s[r]) return solve(l + 1, r, s) || solve(l, r - 1, s);
            l++;
            r--;
        }

        return true;
    }
};